### DISCUSSION TITLE
How to Protect default branch and admin bypass it.

### DISCUSSION THREAD

**Discussion by k-karen** [community/community#142759] [Enterprise] [answered]
I want to setting like following.

Member(= not Admin)
- need 1 approve before merge
- can't push default branch

Admin
- no need approve before merge
- can't push default branch

I did it for that.
- Classic Branch Protection Rule
  - Require a pull request before merging > Require approvals(1 member)
  - (Do not allow bypassing the above settings = false )
- Rule Sets
  - Require a pull request before merging(0 member)

Now, Could someone tell me following?

1. Can these settings be made more better?
2. Can these settings be applied to the entire Organization (Enterprise)?

**Comment by thomassiard** [comment, marked as answer]
1. Refine Classic Branch Protection Rules
Since you’re aiming to control permissions for Members and Admins separately, here’s how you could improve your current setup:

Member Rule:

Set up a branch protection rule for the default branch.
Under Require a pull request before merging, enable Require approvals and set it to 1.
Enable Restrict who can push to matching branches, and select only Admins (or create a custom team with Admin privileges).
Leave Allow bypassing the above settings disabled to ensure Members can't bypass these requirements.
Admin Rule:

Under Rules Sets, you can define a separate rule specifically for Admins if using GitHub’s Enterprise or Team plan.
In Require pull request before merging, set Require approvals to 0 for Admins.
Leave Restrict who can push to matching branches on to prevent Admins from pushing directly to the default branch.
2. Applying Rules Across the Organization or Enterprise
If you have a GitHub Enterprise or GitHub Team account, you can enforce branch protection rules across the organization:

Go to Organization Settings > Code security and analysis.
Under Branch protection rules, create organization-level protection rules that apply to all repositories or certain repository groups.
Apply the above Member and Admin rules at the organization level by creating different branch protection rules for each type of access, ensuring consistency.
3. Improving Settings Further
Consider Code Owners: You could use a CODEOWNERS file for specific directories/files to automatically request reviews from certain members or teams based on the files they work with.
Require Status Checks: For additional security, you can enable Require status checks to pass before merging to ensure that any CI tests are completed successfully before a merge.
Apply Organization-Wide Policies: In GitHub Enterprise, you can enforce organization-wide branch protection and security policies through Enterprise Managed Users (EMU) or Organization Policies.

**Reply by DVSR1411** [reply]
Your point of view is great

**Comment by DVSR1411** [comment]


**Reply by DVSR1411** [reply]
Well @thomassiard is very correct. I can now write your answer as:
- [ ] First, for Members, consider setting up a dedicated branch protection rule specifically for the default branch. This is crucial for maintaining high code quality. One important aspect of this is requiring a pull request before any merging occurs. By enabling the requirement for one approval, you ensure that every change gets a quick review. Additionally, it’s essential to restrict who can push to the matching branches. This should only include Admins or a custom team with Admin privileges, reinforcing control over your codebase. Be sure to disable the option that allows bypassing these settings; this way, Members cannot circumvent these critical checks.
- [ ] On the other hand, for Admins, you can create a separate rule if you’re using GitHub’s Enterprise or Team plan. For Admins, you might allow a quicker process by setting the approval requirement to zero. This means that while they have the flexibility to merge changes without additional approvals, they still can’t push directly to the default branch without going through the established procedures. Keeping those push restrictions in place helps maintain oversight.
- [ ] When it comes to applying these rules across your organization, the good news is that GitHub Enterprise or Team accounts allow you to enforce these rules organization-wide. Start by navigating to your Organization Settings and heading to Code security and analysis. Here, you can create organization-level protection rules that apply to all repositories or specific groups of repositories. This ensures a consistent approach to branch protection, making it clear what everyone should expect.
- [ ] To further enhance your branch protection strategy, you might want to consider utilizing a CODEOWNERS file. This feature automatically requests reviews from specific members or teams based on the files or directories they manage, ensuring that the right people are involved in the review process. Additionally, enabling the requirement for status checks to pass before merging adds another layer of security, ensuring that all automated tests are successful before any changes go live.
- [ ] Lastly, if you’re working within GitHub Enterprise, take advantage of features like Enterprise Managed Users or Organization Policies to enforce branch protection and security standards across your organization. By doing so, you’re not only improving code quality but also fostering a culture of collaboration and accountability within your team.

**Reply by queenofcorgis** [reply]
@DVSR1411 
We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.
